src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +249,140 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+        /// </summary>
+        /// <param name="browserEP">SQL Sever Browser IPEndPoint</param>
+        /// <param name="instanceName">instance name for CLNT_UCAST_INST request</param>
+        /// <returns>SVR_RESP packets from SQL Sever Browser</returns>
+        private static byte[] SendInstanceNameRequest(string browserHostname, string instanceName)
+        {
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
+            }
+
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = 0x04;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
+
+            byte[] responsePacket = null;
+            UdpClient client = null;
+            try
+            {
+                client = new UdpClient(AddressFamily.InterNetwork);
+                Task<int> sendTask = client.SendAsync(requestPacket, requestPacket.Length, browserHostname, SQL_SVR_BROWSER_PORT);
+                Task<UdpReceiveResult> receiveTask = null;
+                if (sendTask.Wait(1000) && (receiveTask = client.ReceiveAsync()).Wait(1000))
+                {
+                    responsePacket = receiveTask.Result.Buffer;
+                }
+            }
+            finally
+            {
+                if (client != null)
+                {
+                    client.Dispose();
+                }
+            }
+
+            return responsePacket;
+        }
 
-            return new SNITCPHandle(serverAndPortParts[0], portNumber, timerExpire, callbackObject, parallel);
+        /// <summary>
+        /// Finds port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <returns>port number for given instance name</returns>
+        private static int GetPortByInstanceName(string browserHostname, string instanceName)
+        {
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                throw new ArgumentException();
+            }
+
+            byte[] responsePacket = SendInstanceNameRequest(browserHostname, instanceName);
+            if (responsePacket == null || responsePacket.Length <= 3 || BytesToInt(responsePacket, 1, 3, true) != responsePacket.Length - 3)
+            {
+                throw new Exception();
+            }
+
+            string serverMessage = Encoding.ASCII.GetString(responsePacket, 3, responsePacket.Length - 3);
+            string[] elements = serverMessage.Split(s_semicolonSeparator);
+            int tcpIndex = Array.IndexOf(elements, "tcp");
```

Comment:
Is it possible that `"tcp"` won't be found in `elements` (i.e. `Array.IndexOf` returns `-1`)?